### PR TITLE
skip additional checks if a merge commit

### PR DESCRIPTION
**Pull request by delboy1978uk**
| Q             | A
| ------------- | ---
| Branch        | master for features and deprecations
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    |no
| Deprecations? | no
| Documented?   | no
| Fixed tickets | https://github.com/phpro/grumphp/issues/1099 

This PR makes Grump skip commit message checks (beyond the defaults) if it is a merge commit. This stops problems like JIRA ticket numbers not being in the main branches like master and develop.

### CODE REVIEW

**Review by oallain — APPROVED** [review]

File: src/Task/Git/CommitMessage.php [hunk 1]
```diff
@@ -135,6 +138,9 @@ public function run(ContextInterface $context): TaskResultInterface
             );
         }
 
+        if ($this->isMergeCommit($commitMessage)) {
```
Comment (veewee) [inline src/Task/Git/CommitMessage.php]: Would it make sense to add a configuration flag like `skip_on_merge_commit` (default `true`) and do the validation at the top of the task - before any other task runs?

That way people could opt-in on checking commit messages on merges (which might be a bit odd, but someone might use it.) Setting this by default to true is a bc break somehow, but I think it makes total sense to do so. People can easily re-activate it by config.

WDYT?

File: src/Task/Git/CommitMessage.php [hunk 2]
```diff
@@ -135,6 +138,9 @@ public function run(ContextInterface $context): TaskResultInterface
...
+            return TaskResult::createSkipped($this, $context);
+        }
 
         if ($this->enforceTypeScopeConventions()) {
```
Comment (veewee) [inline src/Task/Git/CommitMessage.php]: you could do the `isMergeCommit` check for merge commits here and skip the type-scope convetions when it's a merge commit. That could make that conventions logic more straight forward

**Review by veewee — COMMENTED** [review]
Looks like a good starting point. I've added some additional pointers.

File: src/Task/Git/CommitMessage.php (hunk 1 above)
Comment (delboy1978uk) [reply src/Task/Git/CommitMessage.php]: Sounds like a sensible idea, although wouldn't a default of `false` be better, as then no behaviour will change unless you set it. Although as you say, using it on merges is a bit odd and probably the exception.

File: src/Task/Git/CommitMessage.php [hunk 3]
```diff
@@ -56,6 +60,7 @@ public static function getConfigurableOptions(): OptionsResolver
         $resolver->addAllowedTypes('max_body_width', ['int']);
         $resolver->addAllowedTypes('max_subject_width', ['int']);
         $resolver->addAllowedTypes('case_insensitive', ['bool']);
+        $resolver->addAllowedTypes('skip_on_merge_commit', ['bool']);
```
Comment (veewee) [inline src/Task/Git/CommitMessage.php:63]: Can you add this one to the documentation of the task?
https://github.com/phpro/grumphp/blob/v2.x/doc/tasks/git_commit_message.md

File: src/Task/Git/CommitMessage.php [hunk 4]
```diff
@@ -135,8 +144,7 @@ public function run(ContextInterface $context): TaskResultInterface
             );
         }
 
-
-        if ($this->enforceTypeScopeConventions()) {
+        if (!$this->isMergeCommit($commitMessage) && $this->enforceTypeScopeConventions()) {
```
Comment (veewee) [inline src/Task/Git/CommitMessage.php]: instead of checking this twice, you could as well remember it in a var like `$isMergeCommit`

**Review by veewee — COMMENTED** [review]
Looks good. Think this is about all the code it requires, nice! :)
Can you take a look at the comments I made and at the failing testsuite? Looks like a missing type in your function signature

**Comment by veewee** [comment]
Thanks for the addition

**Comment by delboy1978uk** [comment]
@veewee Thanks so much for merging this feature! Can I also create a PR for v1.x ?

**Comment by veewee** [comment]
I wasn't really planning of porting new features back to v1 because of the maintenance overhead.

**Comment by delboy1978uk** [comment]
@veewee Understandable, I'll use a fork of my own until the rest of the servers where I'm working are on PHP 8.1 (which shouldn't be long now). I'm looking forward to the release! Thanks again :D

**Pull request merged**